### ISSUE TITLE
AbstractToJacksonConverter does not work when userType is array

### ISSUE THREAD

**Issue by giger85** [jOOQ/jOOQ#14204]
### Expected behavior

In jooq 3.17, json converter function is supported.

- DB table & DML
```sql
CREATE TABLE bar (
    id              BIGINT                      PRIMARY KEY,
    name            VARCHAR(16)                 NOT NULL,
    json_array_data JSONB                       NULL,
    created_at      TIMESTAMP WITH TIME ZONE    NOT NULL DEFAULT now()
);

INSERT INTO bar (id, name, json_array_data) VALUES (2, 'bar user #2', '[{"key": "firstKey", "value": "firstValue"}]'::jsonb);
```
- gradle code generation config
```groovy
forcedType {
  includeExpression = 'bar.json_array_data'
  includeTypes = 'JSONB'
  userType = 'io.github.giger85.example.BarInfo[]'
  jsonConverter = true
}
```
- deserialize class
```java
public record BarEntity(
        @NotNull Long id,
        @NotNull String name,
        BarInfo[] jsonArrayData,
        @NotNull OffsetDateTime createdAt
) {
}

public record BarInfo(String key, String value) {
}
```

I expected that transforming from `Record` to `BarEntity` will success.

### Actual behavior

If use code generation and set userType as array, then deserialization is fail.

- strace trace
```
org.jooq.exception.DataAccessException: SQL [select "public"."bar"."id", "public"."bar"."name", "public"."bar"."json_array_data", "public"."bar"."created_at" from "public"."bar" where "public"."bar"."id" = ?]; Error while reading field: "public"."bar"."json_array_data", at JDBC index: 3
	at org.jooq_3.17.4.POSTGRES.debug(Unknown Source) ~[na:na]
	at org.jooq.impl.Tools.translate(Tools.java:3307) ~[jooq-3.17.4.jar:na]
	at org.jooq.impl.DefaultExecuteContext.sqlException(DefaultExecuteContext.java:678) ~[jooq-3.17.4.jar:na]
	at org.jooq.impl.CursorImpl$CursorIterator.fetchNext(CursorImpl.java:1464) ~[jooq-3.17.4.jar:na]
	at org.jooq.impl.CursorImpl$CursorIterator.hasNext(CursorImpl.java:1425) ~[jooq-3.17.4.jar:na]
	at org.jooq.impl.CursorImpl.fetchNext(CursorImpl.java:238) ~[jooq-3.17.4.jar:na]
	at org.jooq.impl.Tools.fetchOne(Tools.java:2599) ~[jooq-3.17.4.jar:na]
  ...
Caused by: java.sql.SQLException: Error while reading field: "public"."bar"."json_array_data", at JDBC index: 3
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.setValue(CursorImpl.java:1590) ~[jooq-3.17.4.jar:na]
	...
Caused by: org.jooq.exception.DataTypeException: Error when converting JSON to class io.github.giger85.example.BarInfo
	at org.jooq.jackson.extensions.converters.AbstractToJacksonConverter.from(AbstractToJacksonConverter.java:82) ~[jooq-jackson-extensions-3.17.4.jar:na]
	at org.jooq.jackson.extensions.converters.JSONBtoJacksonConverter.from(JSONBtoJacksonConverter.java:48) ~[jooq-jackson-extensions-3.17.4.jar:na]
	at org.jooq.impl.DefaultBinding$InternalBinding.get(DefaultBinding.java:1025) ~[jooq-3.17.4.jar:na]
	at org.jooq.impl.CursorImpl$CursorIterator$CursorRecordInitialiser.setValue(CursorImpl.java:1580) ~[jooq-3.17.4.jar:na]
	...
Caused by: com.fasterxml.jackson.databind.exc.MismatchedInputException: Cannot deserialize value of type `io.github.giger85.example.BarInfo` from Array value (token `JsonToken.START_ARRAY`)
 at [Source: (String)"[{"key": "firstKey", "value": "firstValue"}]"; line: 1, column: 1]
	at com.fasterxml.jackson.databind.exc.MismatchedInputException.from(MismatchedInputException.java:59) ~[jackson-databind-2.14.0-rc2.jar:2.14.0-rc2]
  ...
```

### Steps to reproduce the problem

In order to reproduce this error, you can execute test cases of below repository.
- https://github.com/giger85/example-jooq/
- prerequisite: docker

```
$ ./gradlew :{cleanTest,test} --tests BarRepositoryIntegrationTests.testFindByIdError --info
```

### jOOQ Version

jOOQ OSS 3.17.4

### Database product and version

PostgreSQL 15.1

### Java Version

openjdk version "17.0.4.1" 2022-08-12

### OS Version

centos 7.9, apple M1

### JDBC driver name and version (include name if unofficial driver)

org.postgresql:postgresql:42.5.0

**Comment by giger85** [comment]
## Additional information
In my opinion, [below code](https://github.com/giger85/example-jooq/blob/61fddb59650759571dad1b881a67514a6574f75b/src/generated/jooq/io/github/giger85/example/tables/Bar.java#L62
) is problem.
```java
/**
* The column <code>public.bar.json_array_data</code>.
*/
public final TableField<Record, BarInfo[]> JSON_ARRAY_DATA = createField(DSL.name("json_array_data"), SQLDataType.JSONB, this, "", new JSONBtoJacksonConverter<BarInfo[]>((Class<BarInfo[]>) (Class) BarInfo.class));
```

When I use fixed converter, deserialization is work for me.
- [test code](https://github.com/giger85/example-jooq/blob/61fddb59650759571dad1b881a67514a6574f75b/src/test/java/io/github/giger85/example/BarRepositoryIntegrationTests.java#L22)
```java
private static final JSONBtoJacksonConverter<BarInfo[]> FIXED_CONVERTER = new JSONBtoJacksonConverter<>(BarInfo[].class);
```

**Comment by lukaseder** [comment]
Thanks a lot for your report. I'll look into this soon

**Comment by lukaseder** [comment]
Looking into this now. I can reproduce the exception from running your reproducer.

**Comment by lukaseder** [comment]
OK, I see, you're expecting to map a SQL `JSONB` type (not `JSONB[]`) onto a Java `T[]` array type. Things seem to work with lists, no?

```patch
diff --git a/build.gradle b/build.gradle
index 3645194..2a85139 100644
--- a/build.gradle
+++ b/build.gradle
@@ -74,7 +74,7 @@ jooq {
                             forcedType {
                                 includeExpression = 'bar.json_array_data'
                                 includeTypes = 'JSONB'
-                                userType = 'io.github.giger85.example.BarInfo[]'
+                                userType = 'java.util.List<io.github.giger85.example.BarInfo>'
                                 jsonConverter = true
                             }
                         }
diff --git a/src/generated/jooq/io/github/giger85/example/tables/Bar.java b/src/generated/jooq/io/github/giger85/example/tables/Bar.java
index 1acdb9d..93c139b 100644
--- a/src/generated/jooq/io/github/giger85/example/tables/Bar.java
+++ b/src/generated/jooq/io/github/giger85/example/tables/Bar.java
@@ -9,6 +9,7 @@ import io.github.giger85.example.Keys;
 import io.github.giger85.example.Public;

 import java.time.OffsetDateTime;
+import java.util.List;

 import org.jooq.Field;
 import org.jooq.ForeignKey;
@@ -59,7 +60,7 @@ public class Bar extends TableImpl<Record> {
     /**
      * The column <code>public.bar.json_array_data</code>.
      */
-    public final TableField<Record, BarInfo[]> JSON_ARRAY_DATA = createField(DSL.name("json_array_data"), SQLDataType.JSONB, this, "", new JSONBtoJacksonConverter<BarInfo[]>((Class<BarInfo[]>) (Class) BarInfo.class));
+    public final TableField<Record, List<BarInfo>> JSON_ARRAY_DATA = createField(DSL.name("json_array_data"), SQLDataType.JSONB, this, "", new JSONBtoJacksonConverter<List<BarInfo>>((Class<List<BarInfo>>) (Class) List.class));

     /**
      * The column <code>public.bar.created_at</code>.
diff --git a/src/main/java/io/github/giger85/example/BarEntity.java b/src/main/java/io/github/giger85/example/BarEntity.java
index ace7f46..20fa436 100644
--- a/src/main/java/io/github/giger85/example/BarEntity.java
+++ b/src/main/java/io/github/giger85/example/BarEntity.java
@@ -4,11 +4,12 @@ package io.github.giger85.example;
 import jakarta.validation.constraints.NotNull;

 import java.time.OffsetDateTime;
+import java.util.List;

 public record BarEntity(
         @NotNull Long id,
         @NotNull String name,
-        BarInfo[] jsonArrayData,
+        List<BarInfo> jsonArrayData,
         @NotNull OffsetDateTime createdAt
 ) {
 }
diff --git a/src/main/java/io/github/giger85/example/DefaultBarRepository.java b/src/main/java/io/github/giger85/example/DefaultBarRepository.java
index c55d29f..10ff795 100644
--- a/src/main/java/io/github/giger85/example/DefaultBarRepository.java
+++ b/src/main/java/io/github/giger85/example/DefaultBarRepository.java
@@ -8,18 +8,19 @@ import org.jooq.jackson.extensions.converters.JSONBtoJacksonConverter;
 import org.springframework.stereotype.Repository;

 import java.time.OffsetDateTime;
+import java.util.List;

 import static io.github.giger85.example.Tables.BAR;
 import static org.jooq.impl.DSL.field;

 @Repository
 public class DefaultBarRepository implements BarRepository {
-    private static final Field<Object> FIXED_JSON_ARRAY_DATA = field("json_array_data");
-    private static final JSONBtoJacksonConverter<BarInfo[]> FIXED_CONVERTER = new JSONBtoJacksonConverter<>(BarInfo[].class);
-    private static final RecordMapper<Record, BarEntity> RECORD_MAPPER = record -> {
+    private static final Field<Object>                          FIXED_JSON_ARRAY_DATA = field("json_array_data");
+    private static final JSONBtoJacksonConverter<List<BarInfo>> FIXED_CONVERTER       = new JSONBtoJacksonConverter<List<BarInfo>>((Class) List.class);
+    private static final RecordMapper<Record, BarEntity>        RECORD_MAPPER         = record -> {
         Long id = record.getValue(BAR.ID);
         String name = record.getValue(BAR.NAME);
-        BarInfo[] jsonArrayData = record.getValue("json_array_data", FIXED_CONVERTER);
+        List<BarInfo> jsonArrayData = record.getValue("json_array_data", FIXED_CONVERTER);
         OffsetDateTime createdAt = record.getValue(BAR.CREATED_AT);

         return new BarEntity(id, name, jsonArrayData, createdAt);
diff --git a/src/test/java/io/github/giger85/example/BarRepositoryIntegrationTests.java b/src/test/java/io/github/giger85/example/BarRepositoryIntegrationTests.java
index 86c3954..3cad751 100644
--- a/src/test/java/io/github/giger85/example/BarRepositoryIntegrationTests.java
+++ b/src/test/java/io/github/giger85/example/BarRepositoryIntegrationTests.java
@@ -29,8 +29,8 @@ class BarRepositoryIntegrationTests {
         assertThat(barEntity.id()).as("Both ids should be same.").isEqualTo(id);
         assertThat(barEntity.name()).isEqualTo("bar user #2");
         assertThat(barEntity.jsonArrayData()).isNotNull();
-        assertThat(barEntity.jsonArrayData()[0].key()).isEqualTo("firstKey");
-        assertThat(barEntity.jsonArrayData()[0].value()).isEqualTo("firstValue");
+        assertThat(barEntity.jsonArrayData().get(0).key()).isEqualTo("firstKey");
+        assertThat(barEntity.jsonArrayData().get(0).value()).isEqualTo("firstValue");
     }

     @Test
```

I'll look into why lists seem to work but not arrays.

**Comment by lukaseder** [comment]
I've tried reproducing this afresh, like this, without any success:

```java
@Test
public void testJSONBConverterWithArrays() {
    JSONBtoJacksonConverter<InformationSchema[]> c = new JSONBtoJacksonConverter<>(InformationSchema[].class);

    // This is a org.jooq.util.xml.jaxb.InformationSchema object with sample content
    InformationSchema is = is();
    JSONB json = c.to(new InformationSchema[] { is, is });

    assertEquals(is, c.from(json)[0]);
    assertEquals(is, c.from(json)[1]);
    assertEquals(json, c.to(c.from(json)));
    assertEquals("[" + jsonString() + "," + jsonString() + "]", json.data());
}
```

**Comment by lukaseder** [comment]
Ah yes, you already analysed this here: https://github.com/jOOQ/jOOQ/issues/14204#issuecomment-1311403467

This does seem to be a code generation problem, not a runtime problem.

**Comment by lukaseder** [comment]
Fixed in jOOQ 3.18.0 and 3.17.6 (https://github.com/jOOQ/jOOQ/issues/14281). Thanks again for your report!

**Comment by giger85** [comment]
@lukaseder 
Thanks for your help.
When `3.17.6` version is released, I will apply this change to my code.
